source/MapPanel.cpp

```diff
@@ -1378,30 +1378,31 @@ void MapPanel::DrawSystems()
 	if(commodity != cachedCommodity)
 		UpdateCache();
 
-	// If coloring by government, we need to keep track of which ones are the
-	// closest to the center of the window because those will be the ones that
-	// are shown in the map key.
+	// If coloring by government, we need to keep track of which ones are
+	// most prevalent.
 	if(commodity == SHOW_GOVERNMENT)
-		closeGovernments.clear();
+		bigGovernments.clear();
 
 	// Draw the circles for the systems.
 	double zoom = Zoom();
+	hasUnexplored = false;
+	hasUninhabited = false;
 	for(const Node &node : nodes)
 	{
 		Point pos = zoom * (node.position + center);
 		RingShader::Draw(pos, OUTER, INNER, node.color);
 
-		if(commodity == SHOW_GOVERNMENT && node.government && node.government->GetName() != "Uninhabited")
+		// Count the system for the government (to chose what goverment labels to display)
+		// Make sure that it is in an oval that is stretched to fit in the centers of all the edges
+		// before counting it
+		if(commodity == SHOW_GOVERNMENT &&
+			(pos * (1 / Screen::BottomRight())).LengthSquared() < 1)
 		{
-			// For every government that is drawn, keep track of how close it
-			// is to the center of the view. The four closest governments
-			// will be displayed in the key.
-			double distance = pos.Length();
-			auto it = closeGovernments.find(node.government);
-			if(it == closeGovernments.end())
-				closeGovernments[node.government] = distance;
-			else
-				it->second = min(it->second, distance);
+			hasUnexplored |= (node.color == UnexploredColor());
+			hasUninhabited |= (node.color == UninhabitedColor());
```

Comment:
What happens if someone creates a government that shares the uninhabited color? The node already has a government field, judging by the old code, so you could just compare that.

Reply:
If the government colors are the same, a label isn't going to do much :P

Reply:
Also, this was to prevent the Drak edge case, so if we did that it wouldn't work if there was another government shaded like uninhabited and had no plants in it. I think the easiest solution is to just say "don't make the government's color uninhabited" on the wiki

Reply:
Or, better yet, a warning on any two government color overlap

Reply:
I don't think we should worry ourselves with that. I'd just consider it a content creator error. If someone chooses confusing colors, that's on them, and the engine shouldn't baby content creators and say "maybe you shouldn't do that." The only time the engine should give warnings or errors is when there's something that's actually done wrong in a way that prevents things from functioning properly. Having two governments share the same color or a government having the same color as uninhabited (or even unexplored) systems isn't functioning improperly, it's functioning exactly as the content creator said it should.

Reply:
I've fixed this, it now has the original criteria (and the count check works properly as well)

Reply:
I've re-opened this with a question: what should happen if there is a government that has uninhabited color, and one of their systems is uninhabited? Should count for uninhabited or for the government? (This is a very minor edge case but will decide how the logic should work)

Reply:
If a system is uninhabited, it should count as uninhabited.

Reply:
Done in [1dacdc8](https://github.com/endless-sky/endless-sky/pull/10140/commits/1dacdc8bfe3df79506d5fd76f46bc1bd09b33c36)